scripts/check_with_errors.R

```diff
@@ -71,3 +71,68 @@ message(n_notes, ' notes found in ', pkg, '.')
 if (log_notes && n_notes > 0) {
     cat(notes, '\n')
 }
+
+
+
+# PEcAn has a lot of legacy code that issues check warnings,
+# such that it's not yet practical to break the build on every warning.
+# Cleaning this up is a long-term goal, but will take time.
+# Meanwhile, we compare against a cached historic check output to enforce that
+# no *new* warnings are added. As historic warnings are removed, we will update
+# the cached results to ensure they stay gone.
+#
+# To compare checks, we take a two-level approach:
+# First by comparing results with rcmdcheck::compare_checks to find any new
+# top-level warnings (e.g. "checking Rd cross-references ... WARNING"),
+# then if those are OK we get fussier and check for new *instances* of existing
+# warnings (e.g. new check increases from 2 bad Rd cross-references to 3).
+
+msg_lines <- function(msg){
+    msg <- strsplit(
+        gsub("\n  ", " ", msg, fixed = TRUE), #leading double-space indicates line wrap
+        split = "\n",
+        fixed = TRUE)
+    msg <- lapply(msg, function(x)x[x != ""])
+    unlist(lapply(msg, function(x)paste(x[[1]], x[-1], sep=": ")))
+}
+
+old_file <- file.path(pkg, "tests", "Rcheck_reference.log")
+
+# To update reference files after fixing an old warning:
+# * Run check_with_errors.R to be sure the check is currently passing
+# * Delete the file you want to update
+# * Uncomment this section, run check_with_errors.R, recomment
+# * Commit updated file
+# if (! file.exists(old_file)) {
+#    cat("No reference check file found. Saving current results as the new standard\n")
+#    cat(chk$stdout, file = old_file)
+#    quit("no")
+# }
+
+old <- rcmdcheck::parse_check(old_file)
+cmp <- rcmdcheck::compare_checks(old, chk)
+
+
+if (cmp$status != "+") {
+    # rcmdcheck found new messages, so check has failed
+    print(cmp)
+    cat("R check of", pkg, "reports the following new problems.",
+        "Please fix these and resubmit:\n")
+    cat(cmp$cmp$output[cmp$cmp$change == 1], sep="\n")
```

Comment:
Any reason not to use `message` here? I guess so this goes to `stdout` instead of `stderr`?

Reply:
Mostly because the existing `check_with_errors` used cat for its (off-by-default) printing and because it was easy to reason about. For Travis purposes I haven't usually worried much about stdout vs stderr, but I could be persuaded.